File: test-suite-kotlin/src/test/kotlin/io/micronaut/docs/server/sse/HeadlineControllerSpec.kt
package io.micronaut.docs.server.sse

import io.kotest.assertions.timing.eventually
import io.kotest.matchers.shouldBe
import io.kotest.core.spec.style.StringSpec
import io.micronaut.context.ApplicationContext
import io.micronaut.http.HttpRequest
import io.micronaut.http.client.sse.SseClient
import io.micronaut.http.sse.Event
import io.micronaut.runtime.server.EmbeddedServer
import org.opentest4j.AssertionFailedError
import reactor.core.publisher.Flux

import java.util.ArrayList
import kotlin.time.DurationUnit
import kotlin.time.ExperimentalTime
import kotlin.time.toDuration

@ExperimentalTime
class HeadlineControllerSpec: StringSpec() {

    val embeddedServer = autoClose(
            ApplicationContext.run(EmbeddedServer::class.java)
    )

    init {
        "test consume eventstream object" {
            val client = embeddedServer.applicationContext.createBean(SseClient::class.java, embeddedServer.url)

            val events = ArrayList<Event<Headline>>()

            Flux.from(client.eventStream(HttpRequest.GET<Any>("/headlines"), Headline::class.java)).subscribe {
                events.add(it)
            }

            eventually(2.toDuration(DurationUnit.SECONDS), AssertionFailedError::class) {
                events.size shouldBe 2
                events[0].data.title shouldBe "Micronaut 1.0 Released"
                events[0].data.description shouldBe "Come and get it"
            }
        }
    }
}

File: test-suite-python/src/test/python/micronaut/docs/server/sse/HeadlineControllerSpec.py
from typing import Annotated

import java
from jakarta.inject import Inject
from micronaut.context.annotation import Property
from micronaut.http import HttpRequest
from micronaut.http.client.annotation import Client
from micronaut.http.client.sse import SseClient
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

Flux = java.type("reactor.core.publisher.Flux")
HeadlineClass = java.type("micronaut.docs.server.sse.Headline")


@Property(name="spec.name", value="HeadlineControllerSpec")
@MicronautTest
class HeadlineControllerSpec:
    client: Annotated[SseClient, Inject, Client("/")]

    @Test
    def testConsumeEventStreamObject(self):
        events = getattr(Flux, "from")(
            self.client.eventStream(HttpRequest.GET("/headlines"), HeadlineClass)
        ).collectList().block()

        assert events.size() == 2, "events size: " + str(events.size())
        assert events.get(0).getData().getTitle() == "Micronaut 1.0 Released"
        assert events.get(0).getData().getDescription() == "Come and get it"
